Fixed #29329 -- Made datetime logging from runserver more consistent.

hramezani opened the pull request:
ticket-29329

hramezani commented:
> Hi @hramezani. Surely this needs a test showing the inconsistency? 🤔

Hi @carltongibson. Thanks for comment. a test for `inconsistency` or `consistency`?
The patch will make the log `consistent`.

carltongibson commented:
A test that would fail on master, but pass with your patch. 

I'm not sure you have the right approach here. Shouldn't we be passing [the `datefmt` `__init__` kwarg](https://github.com/python/cpython/blob/5907e61a8d4da6d0f11bf1062d6d17484560a15e/Lib/logging/__init__.py#L555)?

hramezani commented:
@carltongibson Thanks for the suggestion.
I think the actual bug is about the inconsistency in logs with/without `server_time`.
When a log record has the `server_time`, it will be printed like [here is the test](https://github.com/django/django/blob/8725d04764e3528cbaefd0925501f8403d22ae3e/tests/logging_tests/tests.py#L602:L604). and when the log record doesn't have the `server_time` [format function assign it to record](https://github.com/django/django/blob/8725d04764e3528cbaefd0925501f8403d22ae3e/django/utils/log.py#L189:L190). [here is the test](https://github.com/django/django/blob/master/tests/logging_tests/tests.py#L606:L608)

IMO, we don't need an extra test for it.

BTW, I've created a test to cover it. Don't know the test looks good to you or not.
Please let me know if something needs to change.

carltongibson commented on django/utils/log.py:
```diff
@@ -162,7 +162,7 @@ def filter(self, record):
 class ServerFormatter(logging.Formatter):
     def __init__(self, *args, **kwargs):
         self.style = color_style()
-        super().__init__(*args, **kwargs)
+        super().__init__(*args, datefmt='%d/%b/%Y %H:%M:%S', **kwargs)
```
Are we sure we don't get passed `datefmt` already? (As part of `args`/`kwargs`?)

I think we should have the signature here [match the super class](https://github.com/python/cpython/blob/5907e61a8d4da6d0f11bf1062d6d17484560a15e/Lib/logging/__init__.py#L555). Then provide a default if `datefmt is `None`: 

```
    def __init__(self, fmt=None, datefmt=None, style='%', validate=True)
        self.style = color_style()
        if datefmt is None: 
            datefmt = %d/%b/%Y %H:%M:%S'
        super().__init__(fmt, datefmt, style, validate) 
```

This way we don't crash no matter how we're instantiated, and folks can still customize by passing the parameter. 

(No doubt no one ever does but...) 

What do you think?

hramezani replied:
Thanks for your suggestion. I've changed it a little bit because [`validate` was added in python3.8](https://docs.python.org/3/library/logging.html#logging.Formatter)

carltongibson replied:
Nice work on `validate` 👍

hramezani commented:
Hi @carltongibson.
Thanks for your review.
I've changed the patch with your suggestion(with a little bit change for `validate`) and removed not needed tests.

felixxm commented on django/utils/log.py:
```diff
@@ -160,9 +160,11 @@ def filter(self, record):
-        super().__init__(*args, **kwargs)
+        if datefmt is None:
+            datefmt = '%d/%b/%Y %H:%M:%S'
+        super().__init__(fmt, datefmt, style, **kwargs)
```
I wonder if we can just set `default_time_format` attribute.

felixxm replied:
https://docs.python.org/3.7/library/logging.html#logging.Formatter.formatTime

carltongibson replied:
Yes, that's better.

hramezani replied:
Yes, we can use `default_time_format`. but if we use it the log would be like:
`[02/Apr/2020 03:06:23,590] `
With `,590` as milliseconds at the end. this happens in 
https://github.com/python/cpython/blob/5907e61a8d4da6d0f11bf1062d6d17484560a15e/Lib/logging/__init__.py#L582
and that isn't equal with `WSGIRequestHandler` log which doesn't have millisecond at the end:
`[02/Apr/2020 03:06:23] `.

is there any way to change `default_msec_format` to remove `,miliseconds`?

carltongibson replied:
What happens if that's the empty string `''`?

hramezani replied:
`TypeError: not all arguments converted during string formatting`
Because of the `s = self.default_msec_format % (t, record.msecs)` needs two places for formating. I think `default_msec_format` should have two places for formating.

felixxm replied:
I think it's fine to make them a bit inconsistent (at least for now). I opened an [issue](https://bugs.python.org/issue40300) in Python.
